posts/searching-for-theorems-in-mathlib.md

```diff
@@ -0,0 +1,112 @@
+
+## Tactics
+
+In addition to the search engines and webpages, there are a few tactics you can use to find theorems in mathlib without leaving your Lean development environment.
+
+### `exact?`
+
+The `exact?` tactic is a great way to find theorems in mathlib that finish off a proof goal. At any point in a tactic proof, you can type `exact?` and lean will search through the lemmas you have imported (as well as the file you are working in itself) for a theorem that proves the current goal using assumptions from your context. If it finds one, it will show a helpful message saying `Try this:` which you can click to insert the theorem into your proof.
+
+A good time to use `exact?` is when you have a proof goal that you think is so straightforward that it is likely to be in mathlib already. If `exact?` doesn't find a proof, you can always extract the fact and prove it or `sorry` it as an independent lemma immediately above the theorem you are working on, and `exact?` should start to work. You can then prove the goal yourself and submit the proof to mathlib.
+
+A downside of this approach is that it depends on the *exact* structure of the goal you are trying to prove being present in the library. If you are hoping to prove the goal `a - b = c` and the library only has a form of the theorem that looks like `a = b + c`, `exact?` will not find it. (Helpful in this process is experience with how mathlib prefers theorems to be structured. For example, a form of a theorem with a subtraction is discouraged in favor of a form with an addition, and `>` is almost entirely absent in favor of `<`.)
+
+### `apply?`
+
+The `apply?` tactic is similar to `exact?`. It searches for theorems that can be applied to the current goal, without necessarily requiring that they close the goal (although if they do close the goal, `apply?` will show you that).
+
+`apply?` is useful when the structure of your goal is such that there seems to be an obvious fact to apply, but you don't know the name of the fact. It's pretty much strictly more powerful than `exact?`, but it can be slower to run.
+
+### `rw?`
+
+In the same vein as `exact?` and `apply?`, `rw?` is a tactic that searches for lemmas that can be used to rewrite the current goal. It's useful when you have a subterm of your goal that you know can be rewritten into another form, but you don't know the name of the lemma that does the rewriting. You can also use this to rewrite hypotheses with `rw? at ...`.
+
+### Other Automated Theorem Proving Techniques
+
+There are a few other tactics in mathlib and other projects that don't search for theorems per se, but which try to complete proofs automatically, and which may identify useful theorems in the process. These include:
+
+* `simp`: Simplifies the goal using a set of simplification rules. Using `simp?` allows you to replace `simp` with an equivalent tactic that specifies the simplification rules it uses.
```

Comment:
Can you mention that `simp?` is a good way to learn the lemmas name, that you may then use in a different way if you like (`rw`, proof term, etc.)?